### ISSUE TITLE
Sour Cobalt Tardigrade - Uncontrolled Merkle Root Update cause Unauthorized Minting and Loss of Trust.

### ISSUE THREAD

**Issue by sherlock-admin2** [sherlock-audit/2023-12-footium-judging#68]
Sour Cobalt Tardigrade

high

# Uncontrolled Merkle Root Update cause Unauthorized Minting and Loss of Trust.

## Summary
The `changeMerkleRoot` function allows the owner to update the `merkle` root used for player minting proofs without any safeguards, enabling unauthorized minting by malicious actors. 

## Vulnerability Detail

- Unrestricted Access: Only the owner can call the function, but there are no controls over who proposes or approves updates.
- Direct Assignment: Updating the `merkle` root involves simply assigning the new value to the storage variable without any intermediate checks or confirmation steps.
- Exploitation: An attacker who gains control of the owner account or exploits a vulnerability in the owner contract can update the `merkle` root with a custom one containing proofs for their own unauthorized players.
## Impact
- Unauthorized Minting: Attackers can mint a potentially unlimited number of unauthorized players, diluting the value of existing players and disrupting the game's economy.
- Loss of Trust: Uncontrolled merkle root updates can erode user trust in the game's fairness and security, leading to decreased engagement and potential financial losses.
- Vulnerability Chain Reaction: Gaining control over the merkle root might open doors to further exploits beyond unauthorized minting.
## Code Snippet
https://github.com/sherlock-audit/2023-12-footium/blob/main/footium-eth-shareable/contracts/FootiumAcademy.sol#L68C4-L71C6
## Tool used
- Manual Review

## Test Case:
```solidity
// Gain control of owner account (simulated here)
// ...

// Generate custom merkle root and attack proofs
bytes32 customMerkleRoot = keccak256(// Construct custom merkle tree);
bytes32[] attackProofs = generateProofs(["attackerPlayer"], customMerkleRoot);

// Update merkle root with the custom one
changeMerkleRoot(customMerkleRoot);

// Mint attacker's player using the custom merkle root and proof
mintPlayer(clubId, "attackerPlayer", attackProofs[0]);

```

## logs
- Successful `changeMerkleRoot` event with the attacker's custom `merkle` root.
- Successful `AcademyPlayerMinted` event for the attacker's player.
## Recommendation
- Implement safeguards for `merkle` root updates, such as:
- Timelocks: Introduce a delay between proposing and applying a new `merkle` root, allowing for community or team review and potential objection.
- Community Voting: Implement a decentralized voting system where the community, not just the owner, decides on `merkle` root updates.
- Multi-Signature Approvals: Require approval from multiple authorized individuals before activating a new `merkle` root.

**Comment by thangtranth** [comment]
Escalate

I believe that the issue is valid. 

> From my understanding, this issue could be raised for every ERC721 token, which doesn't make sense 

It is not accurate in my opinion. Only projects that aim for creators to earn royalties from secondary market sales of their NFTs adopt this standard. Projects that don't include a royalty fee typically do not implement it.

> I thought it was a design decision to not have royalties enabled on clubs

The sponsor choose to have royalty charge for the club. The sponsor also confirmed the issue and will fix it. How 
In Opensea, a 2.5% royalty fee is displayed for the collection https://opensea.io/collection/footium-football-clubs 
<img width="345" alt="image" src="https://github.com/sherlock-audit/2023-12-footium-judging/assets/60476800/0bed65f9-4dfb-4e6d-8c01-11173102bb2a">

> Since the game mechanism have changed in which club sales no longer means club owners own players, the previous issue on bypassing of player royalties by selling club as a whole has been fixed.

Club sales with no players does not mean that it is less valuable and royalty can be foregone. Club NFTs are limited, they give users the right to play and can create more player NFTs . Club NFTs are more valuable than player NFTs.

**Comment by sherlock-admin2** [comment]
> Escalate
> 
> I believe that the issue is valid. 
> 
> > From my understanding, this issue could be raised for every ERC721 token, which doesn't make sense 
> 
> It is not accurate in my opinion. Only projects that aim for creators to earn royalties from secondary market sales of their NFTs adopt this standard. Projects that don't include a royalty fee typically do not implement it.
> 
> > I thought it was a design decision to not have royalties enabled on clubs
> 
> The sponsor choose to have royalty charge for the club. The sponsor also confirmed the issue and will fix it. How 
> In Opensea, a 2.5% royalty fee is displayed for the collection https://opensea.io/collection/footium-football-clubs 
> <img width="345" alt="image" src="https://github.com/sherlock-audit/2023-12-footium-judging/assets/60476800/0bed65f9-4dfb-4e6d-8c01-11173102bb2a">
> 
> > Since the game mechanism have changed in which club sales no longer means club owners own players, the previous issue on bypassing of player royalties by selling club as a whole has been fixed.
> 
> Club sales with no players does not mean that it is less valuable and royalty can be foregone. Club NFTs are limited, they give users the right to play and can create more player NFTs . Club NFTs are more valuable than player NFTs.  

You've created a valid escalation!

To remove the escalation from consideration: Delete your comment.

You may delete or edit your escalation comment anytime before the 48-hour escalation window closes. After that, the escalation becomes final.

**Comment by nevillehuang** [comment]
@thangtranth Fair point, I can see this being valid. But is the opensea feature of royalties directly linked to the FootiumClub contracts? If not I am leaning towards agreeing with LSW and @Czar102

**Comment by thangtranth** [comment]
@nevillehuang : Yes, Opensea supports ERC2981. Please find it here: https://support.opensea.io/hc/en-us/articles/14069267158035-How-do-I-set-creator-earnings-on-OpenSea-
<img width="842" alt="image" src="https://github.com/sherlock-audit/2023-12-footium-judging/assets/60476800/35d4d7a0-b4c3-49f7-9118-6e1c3eebb968">

**Comment by nevillehuang** [comment]
Ok I agree with you, I think the details you provided is sufficient to warrant this as medium severity like I initially noted. Will let @Czar102 decide final severity and let @Cergyk add any additional comments.

**Comment by detectiveking123** [comment]
This issue is invalid -- not making the clubs ERC2981 compatible is a design decision. The OpenSea collection you listed doesn't seem like it has been updated to reflect the new contracts (I'm guessing this was from the old version that was fixed). 

@nevillehuang @Czar102 If you read the discussion here (https://github.com/sherlock-audit/2023-04-footium-judging/issues/293) it's clear that the only reason the sponsors wanted to add ERC2981 to clubs was because players were a part of clubs at the time, and they wanted to enforce the royalty on players.

**Comment by thangtranth** [comment]
Not sure why we are trying to guess what the design choice is when our sponsor is already confirmed and will fix the issue.

**Comment by nevillehuang** [comment]
I can confirm, sponsor has confirmed with me that they intend to have royalties on BOTH clubs and players, and this is even more evident with @thangtranth provided info.

**Comment by detectiveking123** [comment]
The sponsors saying something in DMs is not sufficient to validate an issue like this. There must be something in the publicly available documentation, discord, or codebase that suggests that there is a discrepancy between the provided implementation and intended implementation.

**Comment by nevillehuang** [comment]
@detectiveking123 fair point, that is why I pointed out publicly available information mentioned in previous audit contests (which is part of READ.ME) and the opensea royalties feature for footium, which is also publicly accessible info.

Nevertheless, I can see it going both ways, but given sponsor agrees it is a good catch and the-finding has provided value for them, I am incline to keep issue as medium but I’ll leave it in the hands of @Czar102

**Comment by detectiveking123** [comment]
@nevillehuang If you are talking about the sponsor comment on the previous issue, here is exactly what was written:


"To clarify here from the protocol designer, it's correct that we don't intend for clubs to pay for the royalties for all players on transfer. Instead, we anticipate that the club will subjectively be valued in terms of the players that belong to the club. As a result, clubs should implement the EIP2981 standard so that on club sale royalties will be paid back to the protocol developers"

They seem to suggest that ERC2981 should exist on the clubs because players are a part of clubs (and they don't want people to be able to avoid player royalties by selling clubs). Now that players are not a part of clubs, this point is now moot.

**Comment by nevillehuang** [comment]
@detectiveking123 I get your point and can see this issue going both ways

One way of looking at it is the previous comment pointing to a different root cause of bypassing player sale royalties, and the support for royalties was not mentioned explicitly, that is why i initially invalidated it.

Another way of looking at it is the sponsor currently explicitly mentioned they want to support royalties for clubs in current iterations of the contract, even more so when currently existing clubs supports royalties and sponsor confirmed this, albeit after the contest. Even though mechanism of clubs owning players has changed, it is reasonable for watsons  to assume both club sales and player sales should integrate royalties even more so that they now have different valuation.

So I will leave it to @Czar102 to decide final validity but won’t argue the final outcome as I can see both sides of the equation being valid.

**Comment by Czar102** [comment]
I am siding with @detectiveking123 here and I stand by my initial opinion.

Comments on the previous audit and external websites not linked in the README are not sources of truth in an audit. Some information could have been found only by researching the protocol intentions instead of checking them in the available audit resources. (this applies to DMs, too)

But the crux of my point is still the fact that this is a valid design choice. I could literally quote all @detectiveking123's comments to describe my way of thinking about the judgment here.

Hence, planning to leave the issue as is.

**Comment by thangtranth** [comment]
Can you please show me the information in the README that invalidates the issue?

If, even with sponsor confirmation and all publicly provided evidence, it is still deemed insufficient, I won't argue further. As a researcher, in the face of ambiguous situations it's disappointing when the process of seeking truth is discouraged in favor of guessing and assumption.

> I think it is also a design choice 
I thought it was a design decision 
They seem to suggest that

**Comment by 0xllill** [comment]
> If, even with sponsor confirmation and all publicly provided evidence, it is still deemed insufficient, I won't argue further. As a researcher, in the face of ambiguous situations it's disappointing when the process of seeking truth is discouraged in favor of guessing and assumption.

Hello @Czar102 , how can you say it's a design choice when sponsor add a "sponsor confirmed" tag and confirm with lead judge ?

**Comment by Czar102** [comment]
> Can you please show me the information in the README that invalidates the issue?

The information that invalidates the issue is the fact that this is NOT a security issue, but a design decision or an opportunity loss. Hence it's not a loss of funds.

> even with sponsor confirmation and all publicly provided evidence, it is still deemed insufficient

Sponsor confirmation doesn't matter in judging because the judging process only takes into account the context Watsons had when participating in the audit contest. Not sure what do you mean by "publicly provided evidence" here, if it's something external to the audit then it's not evidence here.

Please DON'T argue about the judgment based on the will/won't fix labels as these responses couldn't have been known at the time of the audit and they are used solely during the bug fix phase.

**Comment by 0xllill** [comment]
I totally agree that sponsor shouldn't be taken into account when calculating impact of an issue in most of the cases but when you're argument is only saying that "it's a valid design choice or a design decision" you are assuming something from the sponsor and the fact that the sponsor confirmed it was not should be taken into account to take a fair decision ...

my question still stands : how can you say it's a design choice when sponsor add a "sponsor confirmed" tag and confirmed with lead judge ?

thank you for your time @Czar102

**Comment by Arabadzhiew** [comment]
> if it's something external to the audit then it's not evidence here.

Sorry, but where exactly is that stated in the Sherlock docs? From what I could find from the [Criteria for Issue Validity](https://docs.sherlock.xyz/audits/judging/judging), there is only a rule regarding **Discord messages or DMs** not being considered sources of truth.

Besides that, it's really sad seeing the findings of Watsons that went the extra mile, researched and found **publicly** available information that validates those findings, getting invalidated due to that information not being included in the contest README. After all, aren't we called security **researchers** for a reason?

**Comment by detectiveking123** [comment]
I don't think that NFT listing is even valid publicly available information to be honest. It's unclear if it was updated for the purpose of this audit, or if it was just a fix from the previous audit.

**Comment by shealtielanz** [comment]
@Czar102 
**Sherlocks criteria for validity** states:
> There is a viable scenario (even if unlikely) that could cause the protocol to enter a state where a material amount of funds can be lost. The attack path is possible with assumptions that either mimic on-chain conditions or reflect conditions that have a reasonable chance of becoming true in the future



Also

https://docs.sherlock.xyz/audits/judging/judging#iii.-some-standards-observed
**III. Some standards observed:
Hierarchy of truth** also states:
> Contest README > Sherlock rules for valid issues > Historical decisions.
While considering the validity of an issue in case of any conflict the sources of truth are prioritized in the above order.
For example: In case of conflict between Sherlock rules vs Sherlock's historical decision, Sherlock criteria for issues must be considered the source of truth.
In case of conflict between information in the README vs Sherlock rules, the README overrides Sherlock rules.

My point is the contract here is clearly in scope, the issue here presents a clear loss of funds to the protocol, and it has been previously awarded in the past.
This issue is satisfied by all the branches in the **Hierarchy of truth**. I strongly believe that throwing this issue as Invalid is not seen as fair and opposes Sherlocks criteria for judging issues.

Where The protocol is supposed to earn  $100. but earn $85 and after so many trades the loss increases exponentially, this is a clear loss of funds to the protocol. and as @nevillehuang stated the issue provided value to the sponsors as the acknowledged they lose funds due to they do not implement the standard. 
> **V. How to identify a medium issue:**
..SNIP..
**A material loss of funds, no/minimal profit for the attacker at a considerable cost**

**Comment by Arabadzhiew** [comment]
> I don't think that NFT listing is even valid publicly available information to be honest. It's unclear if it was updated for the purpose of this audit, or if it was just a fix from the previous audit.

Just FYI, the `FootiumClub` NFT contract that is currently used on OpenSea ([0x659cf1306edba213d1fb8f9352b4593a82b05d0c](https://etherscan.io/address/0x659cf1306edba213d1fb8f9352b4593a82b05d0c#code)) is also an implementation that does not have a `FootiumEscrow` tied to it, or in other words, it also does not have club players attached to it (just like the implementation that is in the scope of this audit). The address of the contract is in the listing URL of each NFT https://opensea.io/assets/ethereum/0x659cf1306edba213d1fb8f9352b4593a82b05d0c/63. 

So what exactly is the unclear part here? To me, it's clear that the protocol team intended to use royalties on clubs that don't have players attached to them.

**Comment by detectiveking123** [comment]
That contract is different from the one in the audit though, which further proves my point. How can that OpenSea link be used as evidence for this audit when the set of contracts is different?

Either way, I still believe that this is just a design decision. I will let @Czar102 decide from here.

**Comment by 0xllill** [comment]
> This issue is invalid -- not making the clubs ERC2981 compatible is a design decision. The OpenSea collection you listed doesn't seem like it has been updated to reflect the new contracts (I'm guessing this was from the old version that was fixed).

=> This [version](https://etherscan.io/address/0x659cf1306edba213d1fb8f9352b4593a82b05d0c#code) is updated from the old contest version you are talking about (which had `escrow) so you're guess is false, and it is 95% the same as the contest contract 

> I don't think that NFT listing is even valid publicly available information to be honest. It's unclear if it was updated for the purpose of this audit, or if it was just a fix from the previous audit.

=> How a publicly available information cannot be valid for a contest ? it's like saying a protocol implement uniswap but didn't add deadlines neither slippage and Uniswap documentation saying deadline and slippage is mandatory is not a valid proof that it is needed , if it was unclear you should have asked to sponsor from my point of view 

> They seem to suggest that ERC2981 should exist on the clubs because players are a part of clubs (and they don't want people to be able to avoid player royalties by selling clubs). Now that players are not a part of clubs, this point is now moot.

=> As long as clubs are NFTs they represent money so they still need royalties implemented on them for me

And again sponsors DMs should not represent a point for evaluating impact on a vulnerability but when the only argument is "it's a design decision" and the design come in fact from sponsor, I think sponsor answer should be taken into account for this case, and still how can you say it's a design choice when sponsor add a "sponsor confirmed" tag and confirmed with lead judge ? I don't know

I don't care about the money of this finding or if it is validated or not but if it's not, there should be fair arguments

**Comment by Arabadzhiew** [comment]
> That contract is different from the one in the audit though, which further proves my point. How can that OpenSea link be used as evidence for this audit when the set of contracts is different?
> 
> Either way, I still believe that this is just a design decision. I will let @Czar102 decide from here.

Obviously, the implementation of the contract is different, but that's not the point. The point is that the implementation of the NFT contract used in OpenSea does not have players attached to it, **just like the implementation that is in the scope of this audit**.

And saying that this is a design decision is an **assumption**, while what you are being presented with are **facts**.

**Comment by Czar102** [comment]
Only information available in the audit is taken into account. It hasn't been shown that this design intention was communicated within the audit resources.

> it's like saying a protocol implement uniswap but didn't add deadlines neither slippage and Uniswap documentation saying deadline and slippage is mandatory is not a valid proof that it is needed , if it was unclear you should have asked to sponsor from my point of view

You can use the documentation of a protocol you integrate with to get to know what assumptions you can and can't make. The vulnerability is nevertheless based on internal workings of the integrated protocol (e.g. if the docs states you mustn't do something, but the contracts are immutable and it provably can't go wrong, it wouldn't be an issue). So, external documentations may provide good hints for the audit.

**Comment by Czar102** [comment]
Result:
Invalid
Has duplicates

**Comment by sherlock-admin2** [comment]
Escalations have been resolved successfully!

Escalation status:
- [thangtranth](https://github.com/sherlock-audit/2023-12-footium-judging/issues/68/#issuecomment-1867439248): rejected

**Comment by CergyK** [comment]
Fix LGTM https://github.com/sherlock-audit/2023-12-footium/pull/2
